cloud/aws/actuators/machine/actuator.go

```diff
@@ -66,42 +69,47 @@ type ActuatorParams struct {
 
 // NewActuator returns an actuator.
 func NewActuator(params ActuatorParams) (*Actuator, error) {
+	e, ok := params.EC2Service.(*ec2svc.Service)
+	if !ok {
+		return nil, errors.New("EC2Service is not a real ec2service")
+	}
 	return &Actuator{
 		codec:          params.Codec,
 		ec2:            params.EC2Service,
 		machinesGetter: params.MachinesGetter,
+		infra: &infra.Infra{
+			EC2Svc: e,
+		},
 	}, nil
 }
 
 // Create creates a machine and is invoked by the machine controller.
 func (a *Actuator) Create(cluster *clusterv1.Cluster, machine *clusterv1.Machine) error {
-	// will need this machine config in a bit
-	_, err := a.machineProviderConfig(machine.Spec.ProviderConfig)
+	status, err := a.machineProviderStatus(machine)
 	if err != nil {
-		glog.Errorf("Failed to decode the machine provider config: %v", err)
-		return err
+		return errors.Wrap(err, "failed to get machine provider status")
 	}
-
-	// Get the machine status
-	status, err := a.machineProviderStatus(machine)
+	clusterStatus, err := a.clusterProviderStatus(cluster)
 	if err != nil {
-		return err
+		return errors.Wrap(err, "failed to get cluster provider status")
 	}
-
-	// does the instance exist with a valid status? we're good
-	// otherwise create it and move on.
-	_, err = a.ec2.InstanceIfExists(status.InstanceID)
+	config, err := a.machineProviderConfig(machine)
 	if err != nil {
-		return err
+		return errors.Wrap(err, "failed to get machine config")
 	}
 
-	i, err := a.ec2.CreateInstance(machine)
+	i, err := a.infra.CreateOrGetMachine(machine, status, config, clusterStatus)
 	if err != nil {
-		return err
+		return errors.Wrap(err, "failed to create or get machine")
 	}
 
 	status.InstanceID = &i.ID
 	status.InstanceState = &i.State
+	machine.Status.NodeRef = &corev1.ObjectReference{
+		Kind:      machine.Kind,
+		Namespace: machine.Namespace,
+		Name:      machine.Name,
+	}
```

Comment:
This seems off to me, the NodeRef should link to a Kubernetes `Node` object and not a cluster-api `Machine`. It really only applies to the case of a `Node` that exists in the same cluster as the `Machine` currently.

Reply:
Yeah I was hella confused here

Reply:
At some point it might be nice if we could extend this in a way that the node object could also be enriched with an apiserver endpoint/credentials so that it could live in a remote cluster to be tracked/synced...  Way out of scope for this PR though :rofl: